File: ainpc-core-plugin/src/main/java/ro/ainpc/world/mapping/MappingWandMode.java
package ro.ainpc.world.mapping;

import java.util.Arrays;
import java.util.Locale;
import java.util.Optional;

public enum MappingWandMode {
    REGION("region"),
    PLACE("place"),
    NODE("node"),
    NPC_BIND("npc_bind"),
    QUEST_ANCHOR("quest_anchor");

    private final String id;

    MappingWandMode(String id) {
        this.id = id;
    }

    public String id() {
        return id;
    }

    public MappingDraftKind draftKind() {
        return switch (this) {
            case REGION -> MappingDraftKind.REGION;
            case PLACE -> MappingDraftKind.PLACE;
            case NODE -> MappingDraftKind.NODE;
            case NPC_BIND -> MappingDraftKind.NPC_BIND;
            case QUEST_ANCHOR -> MappingDraftKind.QUEST_ANCHOR;
        };
    }

    public boolean usesPointSelection() {
        return this == NODE || this == NPC_BIND || this == QUEST_ANCHOR;
    }

    public static Optional<MappingWandMode> fromId(String value) {
        if (value == null || value.isBlank()) {
            return Optional.empty();
        }
        String normalized = value.trim().toLowerCase(Locale.ROOT).replace('-', '_');
        return Arrays.stream(values())
            .filter(mode -> mode.id.equalsIgnoreCase(normalized) || mode.name().equalsIgnoreCase(normalized))
            .findFirst();
    }
}

File: ainpc-core-plugin/src/main/kotlin/ro/ainpc/world/mapping/MappingWandMode.kt
package ro.ainpc.world.mapping

import java.util.Locale

enum class MappingWandMode(private val idValue: String) {
    REGION("region"),
    PLACE("place"),
    NODE("node"),
    NPC_BIND("npc_bind"),
    QUEST_ANCHOR("quest_anchor");

    fun id(): String = idValue

    fun draftKind(): MappingDraftKind =
        when (this) {
            REGION -> MappingDraftKind.REGION
            PLACE -> MappingDraftKind.PLACE
            NODE -> MappingDraftKind.NODE
            NPC_BIND -> MappingDraftKind.NPC_BIND
            QUEST_ANCHOR -> MappingDraftKind.QUEST_ANCHOR
        }

    fun usesPointSelection(): Boolean = this == NODE || this == NPC_BIND || this == QUEST_ANCHOR

    companion object {
        @JvmStatic
        fun fromId(value: String?): MappingWandMode? {
            if (value.isNullOrBlank()) {
                return null
            }
            val normalized = value.trim()
                .lowercase(Locale.ROOT)
                .replace('-', '_')
            for (mode in values()) {
                if (mode.idValue.equals(normalized, ignoreCase = true) ||
                    mode.name.equals(normalized, ignoreCase = true)
                ) {
                    return mode
                }
            }
            return null
        }
    }
}
